flypipe/node.py

```diff
@@ -221,51 +222,66 @@ def __call__(self, *args):
         return self.function(*args)
 
     def run(  # pylint: disable=too-many-arguments
-        self,
-        spark=None,
-        parallel=None,
-        inputs=None,
-        pandas_on_spark_use_pandas=False,
-        parameters=None,
+            self,
+            spark: SparkSession = None,
+            parallel: bool = None,
+            inputs: dict = None,
+            pandas_on_spark_use_pandas: bool = False,
+            parameters: dict = None,
+            cache: dict = None,
     ):
+        if spark:
+            if not isinstance(spark, SparkSession):
+                raise TypeError(f"spark is not of type pyspark.sql.SparkSession")
+
         if not inputs:
             inputs = {}
 
-        provided_inputs = {node.key: df for node, df in inputs.items()}
-        self._create_graph(
-            list(provided_inputs.keys()), pandas_on_spark_use_pandas, parameters
-        )
+        run_context = RunContext(spark=spark,
+                                 parallel=parallel,
+                                 provided_inputs=inputs,
+                                 pandas_on_spark_use_pandas=pandas_on_spark_use_pandas,
+                                 parameters=parameters,
+                                 cache_modes=cache)
+
+        self._create_graph(run_context)
 
-        if provided_inputs is None:
-            provided_inputs = {}
+        # re-create graph again with nodes caches
+        run_context.provided_inputs = self.node_graph.load_caches()
         outputs = {
-            key: NodeResult(spark, df, schema=None)
-            for key, df in provided_inputs.items()
+            node.key: NodeResult(spark, df, schema=None)
+            for node, df in run_context.provided_inputs.items()
         }
-        execution_graph = self.node_graph.get_execution_graph()
 
-        if parallel is None:
-            parallel = get_config("default_run_mode") == RunMode.PARALLEL.value
-        if parallel:
-            return self._run_parallel(spark, execution_graph, outputs)
-        return self._run_sequential(spark, execution_graph, outputs)
+        execution_graph = self.node_graph.get_execution_graph(run_context)
+
+        # In case end_node has been given as input, the execution graph will be empty,
+        # in that case return the provided input for the end node
+        if execution_graph.is_empty():
+            end_node_name = self.node_graph.get_end_node_name(self.node_graph.graph)
+            end_node = self.node_graph.get_transformation(end_node_name)
+            return outputs[end_node_name].as_type(end_node.dataframe_type).get_df()

```

Comment:
oh yeah, there can be graph with only 1 node function that returns a pyspark node, however the provided input is given as pandas dataframe, so we have to convert it as the type of the node.

`if execution_graph.is_empty():` changed this condition to check only if end_node_name is in outputs.

```
# In case end_node has been given as input, the execution graph will be empty,
        # in that case return the provided input for the end node
        end_node_name = self.node_graph.get_end_node_name(self.node_graph.graph)
        if end_node_name in outputs:
            end_node = self.node_graph.get_transformation(end_node_name)
            return outputs[end_node_name].as_type(end_node.dataframe_type).get_df()
```

Reply:
Hmmm can you provide a failing scenario for this?